core/src/main/java/org/bitcoinj/evolution/AbstractQuorumState.java

```diff
@@ -0,0 +1,604 @@
+        super(context.getParams());
+        this.context = context;
+        initializeOnce();
+        initialize();
+    }
+
+    public AbstractQuorumState(NetworkParameters params, byte[] payload, int offset) {
+        super(params, payload, offset);
+        initialize();
+    }
+
+    private void initializeOnce() {
+        syncOptions = SimplifiedMasternodeListManager.SyncOptions.SYNC_MINIMUM;
+        syncInterval = 8;
+    }
+
+    private void initialize() {
+        waitingForMNListDiff = false;
+        pendingBlocks = new ArrayList<StoredBlock>();
+        pendingBlocksMap = new LinkedHashMap<Sha256Hash, StoredBlock>();
+    }
+
+    public void setStateManager(QuorumStateManager stateManager) {
+        this.stateManager = stateManager;
+    }
+
+    // TODO: Do we need to keep track of the header chain also?
+    public void setBlockChain(/*AbstractBlockChain headerChain, */AbstractBlockChain blockChain) {
+        this.blockChain = blockChain;
+        //this.headerChain = headerChain;
+        blockStore = blockChain.getBlockStore();
+        //headerStore = headerChain.getBlockStore();
+    }
+
+    public void pushPendingBlock(StoredBlock block) {
+        pendingBlocks.add(block);
+        pendingBlocksMap.put(block.getHeader().getHash(), block);
+    }
+
+    public List<StoredBlock> getPendingBlocks() {
+        return pendingBlocks;
+    }
+
+    public void popPendingBlock() {
+        StoredBlock thisBlock = pendingBlocks.get(0);
+        pendingBlocks.remove(0);
+        pendingBlocksMap.remove(thisBlock.getHeader().getHash());
+    }
+
+    public void clearFailedAttempts() {
+        failedAttempts = 0;
+    }
+
+    public void incrementFailedAttempts() {
+        ++failedAttempts;
+    }
+
+    public int getFailedAttempts() {
+        return failedAttempts;
+    }
+
+    public boolean reachedMaxFailedAttempts() {
+        return failedAttempts > MAX_ATTEMPTS;
+    }
+
+    abstract int getBlockHeightOffset();
+
+    public abstract int getUpdateInterval();
+
+    abstract boolean needsUpdate(StoredBlock nextBlock);
+
+    public abstract void processDiff(@Nullable Peer peer, DiffMessage difference, AbstractBlockChain headersChain, AbstractBlockChain blockChain, boolean isLoadingBootStrap);
+
+    public abstract void requestReset(Peer peer, StoredBlock block);
+
+    public abstract void requestUpdate(Peer peer, StoredBlock block);
+
+    public abstract SimplifiedMasternodeList getListForBlock(Sha256Hash blockHash);
+
+    public abstract SimplifiedMasternodeList getMasternodeList();
+
+    public abstract SimplifiedMasternodeList getMasternodeListAtTip();
+
+    public abstract LinkedHashMap<Sha256Hash, SimplifiedMasternodeList> getMasternodeListCache();
+
+    public abstract LinkedHashMap<Sha256Hash, SimplifiedQuorumList> getQuorumsCache();
+
+    public abstract ArrayList<Masternode> getAllQuorumMembers(LLMQParameters.LLMQType llmqType, Sha256Hash blockHash);
```

Comment:
Here are the interfaces for what will need to be handled by the subclasses that implement DIP4 and DIP24